### DISCUSSION TITLE
Why does cluster_ipv4_cidr has to be part of network_ipv4_cidr?

### DISCUSSION THREAD

**Discussion by schmidp** [kube-hetzner/terraform-hcloud-kube-hetzner#795] [Q&A]
Hi,

I configured our cluster with:

  network_ipv4_cidr = "172.21.0.0/16"
  cluster_ipv4_cidr = "10.42.0.0/16"

and everything seems to work fine.
Now I noticed that it says:
  # Cluster CIDR must be a part of the network CIDR!

What is the reason for this?
Or what will break if I keep my cluster as is?

All the best, Philipp

**Comment by mysticaltech** [comment]
@schmidp The cluster CIDR is built on top of the Hetzner private network, and routes are assigned between nodes, cluster CIDR and network CIDR. Hence, it's important they share a common subset at least.

Here's what GPT-4 has to say about it:

The network CIDR (Classless Inter-Domain Routing) block represents the entire pool of IP addresses available within your network. The cluster CIDR block, on the other hand, is a subset of IP addresses reserved specifically for your Kubernetes cluster (for Pods and Services, for instance).

Generally, it is advised to have the cluster CIDR as a subset of the network CIDR to ensure that there's no overlap with other parts of your network. Overlapping can lead to routing issues, where traffic might get sent to the wrong place because the same IP range is being used in two different areas.

In your case, the cluster CIDR ("10.42.0.0/16") and network CIDR ("172.21.0.0/16") are distinct and do not overlap. This setup could work without issues if you manage the IP ranges carefully and avoid any overlap.

However, it could potentially cause confusion or misconfiguration in the future. For example, if you or someone else on your team forgets about this setup and deploys something in the "10.42.0.0/16" range outside of the Kubernetes cluster, there could be routing issues or conflicts.

If there is no particular reason to keep these CIDRs separate, I would suggest adjusting them to ensure that the cluster CIDR is a subset of the network CIDR for better network management and to reduce the chance of future issues.
